### PR TITLE
7013 add types

### PR DESCRIPTION
**Pull request by mehdi-aouadi**
## PR Description
Add signed block and signed block contents deserialisation types

[Spec reference here](https://github.com/ethereum/beacon-APIs/pull/302)

## Fixed Issue(s)

#7013 

## Documentation

- [ ] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [ ] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

**Comment by mehdi-aouadi** [comment]
PS: This PR is rebased on https://github.com/ConsenSys/teku/pull/7145

File: data/serializer/src/main/java/tech/pegasys/teku/api/schema/BeaconBlockBody.java [hunk 1]
```diff
@@ -25,25 +32,75 @@
...
+              "deposits",
+              new DeserializableListTypeDefinition<>(DEPOSIT_TYPE),
+              BeaconBlockBody::getDeposits,
+              BeaconBlockBody::setDeposits)
+          .withField(
+              "voluntary_exits",
+              new DeserializableListTypeDefinition<>(SIGNED_VOLUNTARY_EXIT_TYPE),
+              BeaconBlockBody::getVoluntaryExits,
+              BeaconBlockBody::setVoluntaryExits)
+          .build();
```
Comment (rolfyone) [inline data/serializer/src/main/java/tech/pegasys/teku/api/schema/BeaconBlockBody.java]: we should be able to get object type defs from any SSZ object?

also they definitely shouldn't be in 'schema' - the 'schema' objects are the non typedef...

File: data/serializer/src/main/java/tech/pegasys/teku/api/schema/BeaconBlockHeader.java [hunk 2]
```diff
@@ -21,24 +21,56 @@
...
+              CoreTypes.BYTES32_TYPE,
+              BeaconBlockHeader::getStateRoot,
+              BeaconBlockHeader::setStateRoot)
+          .withField(
+              "body_root",
+              CoreTypes.BYTES32_TYPE,
+              BeaconBlockHeader::getBodyRoot,
+              BeaconBlockHeader::setBodyRoot)
+          .build();
+
```
Comment (rolfyone) [inline data/serializer/src/main/java/tech/pegasys/teku/api/schema/BeaconBlockHeader.java]: same as other shcema object comment really...

File: data/serializer/src/main/java/tech/pegasys/teku/api/schema/Checkpoint.java [hunk 3]
```diff
@@ -21,16 +21,27 @@
...
   @Schema(type = "string", format = "byte", description = DESCRIPTION_BYTES32)
-  public final Bytes32 root;
+  public Bytes32 root;
+
+  public static final DeserializableTypeDefinition<Checkpoint> CHECKPOINT_TYPE =
+      DeserializableTypeDefinition.object(Checkpoint.class)
+          .initializer(Checkpoint::new)
+          .withField("epoch", CoreTypes.UINT64_TYPE, Checkpoint::getEpoch, Checkpoint::setEpoch)
+          .withField("root", CoreTypes.BYTES32_TYPE, Checkpoint::getRoot, Checkpoint::setRoot)
+          .build();
```
Comment (rolfyone) [inline data/serializer/src/main/java/tech/pegasys/teku/api/schema/Checkpoint.java]: same - schema object with type defs

File: data/serializer/src/main/java/tech/pegasys/teku/api/schema/Deposit.java [hunk 4]
```diff
@@ -23,13 +24,29 @@
...
+  public static final DeserializableTypeDefinition<Deposit> DEPOSIT_TYPE =
+      DeserializableTypeDefinition.object(Deposit.class)
+          .initializer(Deposit::new)
+          .withField(
+              "proof",
+              new DeserializableListTypeDefinition<>(CoreTypes.BYTES32_TYPE),
+              Deposit::getProof,
+              Deposit::setProof)
+          .withField("data", DEPOSIT_DATA_TYPE, Deposit::getData, Deposit::setData)
+          .build();
```
Comment (rolfyone) [inline data/serializer/src/main/java/tech/pegasys/teku/api/schema/Deposit.java]: same

File: data/serializer/src/main/java/tech/pegasys/teku/api/schema/DepositData.java [hunk 5]
```diff
@@ -22,21 +24,40 @@
...
+          .withField(
+              "withdrawal_credentials",
+              CoreTypes.BYTES32_TYPE,
+              DepositData::getWithdrawalCredentials,
+              DepositData::setWithdrawalCredentials)
+          .withField(
+              "amount", CoreTypes.UINT64_TYPE, DepositData::getAmount, DepositData::setAmount)
+          .withField(
+              "signature", BLS_SIGNATURE_TYPE, DepositData::getSignature, DepositData::setSignature)
+          .build();
```
Comment (rolfyone) [inline data/serializer/src/main/java/tech/pegasys/teku/api/schema/DepositData.java]: same

File: data/serializer/src/main/java/tech/pegasys/teku/api/schema/Eth1Data.java [hunk 6]
```diff
@@ -20,18 +20,39 @@
...
+              Eth1Data::getDepositRoot,
+              Eth1Data::setDepositRoot)
+          .withField(
+              "deposit_count",
+              CoreTypes.UINT64_TYPE,
+              Eth1Data::getDepositCount,
+              Eth1Data::setDepositCount)
+          .withField(
+              "block_hash", CoreTypes.BYTES32_TYPE, Eth1Data::getBlockHash, Eth1Data::setBlockHash)
+          .build();
```
Comment (rolfyone) [inline data/serializer/src/main/java/tech/pegasys/teku/api/schema/Eth1Data.java]: same

File: data/serializer/src/main/java/tech/pegasys/teku/api/schema/IndexedAttestation.java [hunk 7]
```diff
@@ -30,12 +35,34 @@
...
+          .withField(
+              "data",
+              ATTESTATION_DATA_TYPE,
+              IndexedAttestation::getData,
+              IndexedAttestation::setData)
+          .withField(
+              "signature",
+              BLS_SIGNATURE_TYPE,
+              IndexedAttestation::getSignature,
+              IndexedAttestation::setSignature)
```
Comment (rolfyone) [inline data/serializer/src/main/java/tech/pegasys/teku/api/schema/IndexedAttestation.java]: same

File: data/serializer/src/main/java/tech/pegasys/teku/api/schema/KZGCommitment.java [hunk 8]
```diff
@@ -17,13 +17,21 @@
...
   private static final int SIZE = 48;
 
   private final Bytes bytes;
 
+  public static final DeserializableTypeDefinition<KZGCommitment> KZG_COMMITMENT_TYPE =
+      DeserializableTypeDefinition.string(KZGCommitment.class)
+          .formatter(KZGCommitment::toHexString)
+          .parser(KZGCommitment::fromHexString)
+          .format("byte")
+          .build();
```
Comment (rolfyone) [inline data/serializer/src/main/java/tech/pegasys/teku/api/schema/KZGCommitment.java]: same - if this is ssz, shouldnt need to do much

File: data/serializer/src/main/java/tech/pegasys/teku/api/schema/BeaconBlockBody.java (hunk 1 above)
Comment (mehdi-aouadi) [reply data/serializer/src/main/java/tech/pegasys/teku/api/schema/BeaconBlockBody.java]: Fixed

**Comment by mehdi-aouadi** [comment]
@rolfyone Cleaned as discussed. Looks way better now thanks!

**Review by rolfyone — APPROVED** [review]
LGTM

**Pull request merged**